packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -49,50 +57,70 @@ class ProtocolDiscovery {
   final int devicePort;
   final bool ipv6;
 
-  final Completer<Uri> _completer = Completer<Uri>();
+  /// The time to wait before forwarding a new observatory URIs from [logReader].
+  final Duration throttleDuration;
 
   StreamSubscription<String> _deviceLogSubscription;
+  StreamController<Uri> _uriStreamController;
 
   /// The discovered service URI.
+  /// Use [uris] instead.
+  // TODO(egarciad): replace `uri` for `uris`.
+  Future<Uri> get uri {
+    return uris.first;
+  }
+
+  /// The discovered service URI stream.
   ///
   /// Port forwarding is only attempted when this is invoked, in case we never
   /// need to port forward.
-  Future<Uri> get uri async {
-    final Uri rawUri = await _completer.future;
-    return await _forwardPort(rawUri);
+  ///
+  /// Dependending on the lifespan of the app running the observatory,
+  /// a new observatory URI may be assigned to the app.
+  Stream<Uri> get uris {
+    return _uriStreamController.stream
+      .transform(_throttle<Uri>(
+        waitDuration: throttleDuration,
+      ))
+      .asyncMap<Uri>((Uri observatoryUri) => _forwardPort(observatoryUri));
   }
 
   Future<void> cancel() => _stopScrapingLogs();
 
   Future<void> _stopScrapingLogs() async {
+    await _uriStreamController?.close();
     await _deviceLogSubscription?.cancel();
     _deviceLogSubscription = null;
   }
 
-  void _handleLine(String line) {
-    Uri uri;
+  Match _getPatternMatch(String line) {
     final RegExp r = RegExp('${RegExp.escape(serviceName)} listening on ((http|\/\/)[a-zA-Z0-9:/=_\\-\.\\[\\]]+)');
-    final Match match = r.firstMatch(line);
+    return r.firstMatch(line);
+  }
 
+  Uri _getObservatoryUri(String line) {
+    final Match match = _getPatternMatch(line);
     if (match != null) {
-      try {
-        uri = Uri.parse(match[1]);
-      } on FormatException catch (error, stackTrace) {
-        _stopScrapingLogs();
-        _completer.completeError(error, stackTrace);
-      }
+      return Uri.parse(match[1]);
```

Comment:
Instead of catching the FormatError below, you can get the same behavior as the non-matching case by using `Uri.tryParse` which returns `null` on failures

Reply:
Currently, the `Future` completes with the associated error object and stack trace. It might be the case that this isn't needed anymore, but I'd prefer to make behavioral changes like this in a separate PR.